### PR TITLE
Submit Releases/1.0 (resolves #155, resolves #156, resolves #200)

### LINKED ISSUE #155: Fix BWA parsing
Currently doesn't check if a line is empty.

### LINKED ISSUE #156: Cleanup BWA launch script
mesos launch script should not have `--output_dir` set by default, especially writing to `/home/mesosbox/shared`.

### LINKED ISSUE #200: Releases 1.0
Test production pipelines
- [x] CGL
- [x] UNC
- [x] BWA
- [x] Exome

### PR DESCRIPTION
**Pull request by jvivian**
CGCloud version: 1.4a1.dev0
Toil-box AMI built on: 2016-03-22

RNA-seq CGL Pipeline
s3://cgl-driver-projects/test/test_SRR1559177_Sample_NB2318tumor1_T_C3WFVACXX.tar.gz

RNA-seq UNC Pipeline
s3://cgl-driver-projects/test/test_unc_SRR1559177_Sample_NB2318tumor1_T_C3WFVACXX.tar.gz

BWA Alignment Pipeline
s3://cgl-driver-projects/test/alignment/test_align_CPBT_0008_1_tumor.bam

Exome variant pipeline
s3://cgl-driver-projects/test/variants/DTB-005.tar.gz

### CODE REVIEW

File: src/toil_scripts/batch_alignment/bwa_alignment.py [hunk 1]
```diff
@@ -258,6 +256,23 @@ def tarball_files(work_dir, tar_name, uuid=None, files=None):
                 f_out.add(os.path.join(work_dir, fname), arcname=fname)
 
 
+def download_from_s3_url(file_path, url):
```
Comment (fnothaft) [inline src/toil_scripts/batch_alignment/bwa_alignment.py]: Why is this code coming back in here, as opposed to using the library function?

File: src/toil_scripts/batch_alignment/bwa_alignment.py [hunk 2]
```diff
@@ -298,10 +313,12 @@ def parse_config(job, shared_ids, input_args):
 
     with open(config, 'r') as f_in:
         for line in f_in:
-            line = line.strip().split(',')
-            uuid = line[0]
-            urls = line[1:]
-            samples.append((uuid, urls))
+            if line.strip():
+                line = line.strip().split(',')
+                assert len(line) == 3, 'Improper formatting'
```
Comment (fnothaft) [inline src/toil_scripts/batch_alignment/bwa_alignment.py]: @hannes-ucsc and I argued a bit about this a while ago, but I prefer raising named exceptions when validating user input (e.g., if a user provides bad arguments, throw a `IllegalArgumentException` as a Java equivalent) and asserts for checking program invariants (i.e., if the assert failed, there is a bug in the code somewhere, not bad user input).

Also, I'd prefer a more detailed error message, e.g., `Improper formatting. Expected 3 fields, received...`

File: src/toil_scripts/batch_alignment/bwa_alignment.py [hunk 3]
```diff
@@ -576,26 +593,25 @@ def upload_to_s3(work_dir, input_args, output_file):
...
-                       'upload',
-                       '--sse-key-file', os.path.join(work_dir, uuid + '.key'),
-                       '--part-size=64M',
-                       '--upload-slots={}'.format(cores),
-                       '--download-slots={}'.format(cores),
-                       'file://{}'.format(os.path.join(work_dir, output_file)),
-                       s3_path]
+        key_path = input_args['ssec']
 
+        base_url = 'https://s3.amazonaws.com/'
```
Comment (fnothaft) [inline src/toil_scripts/batch_alignment/bwa_alignment.py]: This is `us-east-1` only, right?

File: src/toil_scripts/batch_alignment/launch_bwa_hg19.sh [hunk 4]
```diff
@@ -23,4 +23,4 @@ ${HOME}/toil_mnt/jobStore \
 --output_dir ${HOME} \
 --s3_dir cgl-driver-projects/test/alignment \
 --sudo \
-#--restart
\ No newline at end of file
+#--restart
```
Comment (fnothaft) [inline src/toil_scripts/batch_alignment/launch_bwa_hg19.sh]: Nit: Remove dead args?

File: src/toil_scripts/batch_alignment/launch_bwa_hg19_mesos.sh [hunk 5]
```diff
@@ -22,4 +22,4 @@ aws:us-west-2:alignment-wcdt-run-1 \
 --batchSystem="mesos" \
 --mesosMaster mesos-master:5050 \
 --workDir=/var/lib/toil \
-#--restart
\ No newline at end of file
+#--restart
```
Comment (fnothaft) [inline src/toil_scripts/batch_alignment/launch_bwa_hg19_mesos.sh]: Nit: Remove dead args?

File: src/toil_scripts/batch_alignment/launch_bwa_hg38.sh [hunk 6]
```diff
@@ -23,4 +23,4 @@ ${HOME}/toil_mnt/jobStore \
 --output_dir ${HOME} \
 --s3_dir cgl-driver-projects/test/alignment \
 --sudo \
-#--restart
\ No newline at end of file
+#--restart
```
Comment (fnothaft) [inline src/toil_scripts/batch_alignment/launch_bwa_hg38.sh]: Nit: Remove dead args?

**Comment by fnothaft** [comment]
I just made a pass over the BWA code, since we depend on that in the ADAM/GATK pipeline. Why is the `download_from_s3` code migrating out of the library and back into the scripts?

**Comment by jvivian** [comment]
This PR is just supposed to satisfy the production team with working toil scripts so development can move forward. Most of the code really needs to be rewritten and I had to stop myself from over-refactoring. 

> Why is this code coming back in here, as opposed to using the library function?

The releases/1.0 branched before the common libraries PR and traditionally our production team was familiar with running pipelines from a single script.  Once common libraries has been merged, I'm going to refactor all of these to use the common library functions wherever possible. 

> but I prefer raising named exceptions when validating user input (e.g., if a user provides bad arguments, throw a IllegalArgumentException as a Java equivalent) and asserts for checking program invariants (i.e., if the assert failed, there is a bug in the code somewhere, not bad user input).

I like that and will implement it in the future.

> This is us-east-1 only, right?

Good catch, i'll fix for this PR. 

> Nit: Remove dead args?

When I was first writing these, this was sort of a hint to let people know `--restart` was an option, but I agree that it should be removed.

**Comment by fnothaft** [comment]
> This PR is just supposed to satisfy the production team with working toil scripts so development can move forward. Most of the code really needs to be rewritten and I had to stop myself from over-refactoring.

SGTM; thanks for explaining the context, I won't get in the way of a quick merge.

RE: refactoring, perhaps the two of us (+ any interested parties) should do a non-scrum call to discuss a refactor soon? There are lots of touchpoints between our work, e.g., the batch alignment and preprocessing pipelines.

> The releases/1.0 branched before the common libraries PR and traditionally our production team was familiar with running pipelines from a single script. Once common libraries has been merged, I'm going to refactor all of these to use the common library functions wherever possible.

So would this merge before #190, and then #190 will factor the calls out? I've been meaning to go over #190; I'll make sure to make a pass today.

**Comment by jvivian** [comment]
> refactoring, perhaps the two of us (+ any interested parties) should do a non-scrum call to discuss a refactor soon?

Ya, that sounds good to me — I have a couple pipelines based off the #190 branch (no open PRs for them atm), that are much cleaner than the old pipelines (with the side effect of making me cringe anytime i have to look at the old ones). 

> So would this merge before #190, and then #190 will factor the calls out? I've been meaning to go over #190; I'll make sure to make a pass today.

This will ideally merge today and we'll tag it as the official production release to use for the next month or so.  #190 will only add the common libraries. I feel like after #190 is merged, individual tickets should be assigned to refactor the individual pipelines.   

>  I've been meaning to go over #190; I'll make sure to make a pass today.

That'd be great — Hannes has been very thorough about reviewing the commons PR, but we'd both appreciate an extra set of eyes on it.

File: src/toil_scripts/batch_alignment/bwa_alignment.py (hunk 2 above)
Comment (hannes-ucsc) [inline src/toil_scripts/batch_alignment/bwa_alignment.py]: Agreed.

File: src/toil_scripts/batch_alignment/bwa_alignment.py (hunk 1 above)
Comment (hannes-ucsc) [inline src/toil_scripts/batch_alignment/bwa_alignment.py]: We determined that the "common library" PR will not be included in the 1.0 release.

**Comment by fnothaft** [comment]
> Ya, that sounds good to me — I have a couple pipelines based off the #190 branch (no open PRs for them atm), that are much cleaner than the old pipelines (with the side effect of making me cringe anytime i have to look at the old ones).

Yeah, @heuermh has some code that he was saying will be much cleaner post #190.

> This will ideally merge today and we'll tag it as the official production release to use for the next month or so. #190 will only add the common libraries. I feel like after #190 is merged, individual tickets should be assigned to refactor the individual pipelines.

SGTM. Also, moving forward, perhaps we should make better use of the milestone tags. At least where releases are concerned, they're a good way to signal what is targeted for a release and what isn't.

Also, do you all have a writeup for your repository branch/release model? I had seen the `releases/1.0` branch floating around for a while. Clearly we're not doing gitflow, but that just means that we could be following one of any other millions of release models.

Additionally, it might be good to drop documentation in that clearly indicates _which_ pipelines are at 1.0. I was a bit surprised to see a 1.0 tag get proposed because I work so much on the ADAM/GATK stuff, which is... alpha or maybe beta. We should probably make a general documentation writing push after both papers are out the door.

**Comment by jvivian** [comment]
@hannes-ucsc  is a fan of Gitflow and I believe our intention is to move to that model.

**Comment by fnothaft** [comment]
> @hannes-ucsc is a fan of Gitflow and I believe our intention is to move to that model.

Ah, OK, cool. Since it'd be a switch in development methodology (I mean, if we oversimplify, it's just a change in how we open PRs and various conventions around branching), what's the plan for cutting over? E.g., timeline, etc.

**Comment by hannes-ucsc** [comment]
I do like how gitflow unblocks new development on the develop branch by forking off a release branch. I don't like how gitflow only allows a linear sequence of releases. IOW, AFAIK, I can't support two stable releases, which for Toil I plan on doing. I also don't like that I have to merge a release branch into master *before* I can build off of it.

I'm not sure I want to switch completely to gitflow/hubflow. I don't care about the hubflow tooling because I have SourceTree but @briandoconnor says it eases the learning curve. Maybe we can use the develop/feature branch side of gitflow and let people optionally use the hubflow tooling. But for the release/hotfix side of things use my flow.

**Comment by fnothaft** [comment]
> I do like how gitflow unblocks new development on the develop branch by forking off a release branch. 

Agreed; that's a nice property.

> I don't like how gitflow only allows a linear sequence of releases. IOW, AFAIK, I can't support two stable releases, which for Toil I plan on doing.

Is the use case here that you want to propagate hotfixes to point releases based off of old releases? I feel like that would be a straightforward extension of gitflow (should just change where hotfix branches can merge), if not strict gitflow.

> I also don't like that I have to merge a release branch into master before I can build off of it.

Agreed; that is clunky.

**Comment by jvivian** [comment]
@briandoconnor asked if this would be ready by Friday — should i squash so it can be merged or are there outstanding issues?

**Comment by fnothaft** [comment]
> @briandoconnor asked if this would be ready by Friday — should i squash so it can be merged or are there outstanding issues?

Don't know if my opinion counts for much on this PR, but LGTM! ;)

**Comment by hannes-ucsc** [comment]
Outstanding issue moved to 

https://github.com/BD2KGenomics/toil-scripts/issues/211

Please squash and force push.

**Comment by jvivian** [comment]
> Please squash and force push.

Done.

@hannes-ucsc

**Pull request merged**